### PR TITLE
contrib: verify torrents with verify-binary.py

### LINKED ISSUE #27702: Include torrent in binary download verification
The binary verification script currently covers bitcoin(core).org but not the torrents. We should add an option to download and verify those as well.

### PR DESCRIPTION
**Pull request by willcl-ark**
Closes #27702

Adds functionality to the verify-binary.py script to verify the signatures in a torrent download.
Users must download the torrent themselves manually and point the script to the directory using the `torrent` subcommand.

The syntax chosen was `./contrib/verify-binaries/verify.py torrent <path to SHA256SUMS file>`

### CODE REVIEW

File: contrib/verify-binaries/verify.py [hunk 1]
```diff
@@ -459,14 +460,17 @@ def verify_binary_hashes(hashes_to_verify: t.List[t.List[str]]) -> t.Tuple[Retur
...
 
 def print_output(
-    args: argparse.Namespace,
-    good_trusted: t.List[SigData],
-    good_untrusted: t.List[SigData],
-    unknown: t.List[SigData],
-    bad: t.List[SigData],
-    files_to_hashes: t.Dict[str, str],
-    missing_files=None,
-    ):
```
Comment (Sjors) [inline contrib/verify-binaries/verify.py]: a6dcf2cb0817143beacf0b9004191a0dd6d2e43c Did you mean to change the indentation here? Maybe do so in the first refactor commit?

File: contrib/verify-binaries/verify.py [hunk 2]
```diff
@@ -481,9 +485,8 @@ def print_output(
     else:
         for filename in files_to_hashes:
             print(f"VERIFIED: {filename}")
-        if missing_files:
-            for filename in missing_files:
-                print(f"MISSING: {filename}")
+        for filename in missing_files:
```
Comment (Sjors) [inline contrib/verify-binaries/verify.py:488]: a6dcf2cb0817143beacf0b9004191a0dd6d2e43c: did you mean to do this in the refactor commit?

**Review by Sjors — COMMENTED** [review]
Successfully tested on macOS 13.4 against v25.0.

Lightly code reviewed a6dcf2cb0817143beacf0b9004191a0dd6d2e43c.

Do we want to use the .asc file from the torrent, or should we just get it from the Guix signature repo? The latter may have more signatures, as they keep coming in after release.

File: contrib/verify-binaries/verify.py (hunk 1 above)
Comment (willcl-ark) [reply contrib/verify-binaries/verify.py]: Thanks fixed in next push

File: contrib/verify-binaries/verify.py (hunk 2 above)
Comment (willcl-ark) [reply contrib/verify-binaries/verify.py:488]: Thanks fixed in next push

**Comment by kristapsk** [comment]
> Do we want to use the .asc file from the torrent, or should we just get it from the Guix signature repo? The latter may have more signatures, as they keep coming in after release.

It's pros and cons. Repo will have more signatures, but any centralized server is single point of failure, torrents are better in this regard.

**Comment by willcl-ark** [comment]
Sure there are tradeoffs.

It also seems problematic in that fetching them in any sane way would  require `git` to be installed on the host machine (so we could e.g. do a sparse checkout of the release sub-directory). Alternatively just download the entire repo as .zip and go from there? Both seem undesirable/"messy" to me personally. 🤷🏼‍♂️

**Comment by Sjors** [comment]
In the case of Github, I believe you can get individual raw files too.

But there's something to be said for not being dependent on having Git installed, and also have the option to not download if you used a torrent before. So maybe we can have a `--signatures-from-repo` like option in that case? For non-torrent imo this should be `true` by default if git is installed. For torrents the option could be suggested.

**Comment by willcl-ark** [comment]
OK I implemented as you suggested @Sjors. It does seem like a win to get more signatures on balance.

If people feel the approach is OK, and that it belongs in this PR (and not it's own), I can probably tidy it up further as theres still some duplicated code which can probably be factored out.

File: contrib/verify-binaries/verify.py [hunk 3]
```diff
@@ -692,16 +866,23 @@ def main():
...
     bin_parser.set_defaults(func=verify_binaries_handler)
     bin_parser.add_argument("--sums-sig-file", "-s", help="Path to the SHA256SUMS.asc file to verify")
+    bin_parser.add_argument("--signatures-from-repo", action="store", default=True, type=arg_str_to_bool, help="Fetch additional signatures from the github repo. Requires git to be installed")
     bin_parser.add_argument("sums_file", help="Path to the SHA256SUMS file to verify")
     bin_parser.add_argument(
         "binary", nargs="*",
         help="Path to a binary distribution file to verify. Can be specified multiple times for multiple files to verify."
     )
 
+    torrent_parser = subparsers.add_parser("torrent", help="Verify local torrent dir (requires manual torrent download).")
```
[the remark this answers is not in the document]
Comment (willcl-ark) [reply contrib/verify-binaries/verify.py:881]: Right, without performing the actual torrent download it ends up being the same!

I'm going to close this for now as IMO it doesn't make sense to pull in a large python torrent lib just to perform the download in this verification script. Perhaps we close #27702 as well, unless we really want this.

If there was appetite to add a python-based torrent download, I'd be happy to work on it...

**Pull request closed without merging**